tests/phpunit/test-class-wc-stripe-upe-payment-gateway.php

```diff
@@ -0,0 +1,556 @@
+
+		$payment_intent_mock                       = self::MOCK_CARD_PAYMENT_INTENT_TEMPLATE;
+		$payment_intent_mock['id']                 = $payment_intent_id;
+		$payment_intent_mock['amount']             = $amount;
+		$payment_intent_mock['last_payment_error'] = [];
+		$payment_intent_mock['payment_method']     = $payment_method_mock;
+		$payment_intent_mock['charges']['data'][0]['payment_method_details'] = $payment_method_mock;
+
+		$this->mock_gateway->expects( $this->once() )
+			->method( 'stripe_request' )
+			->with( "payment_intents/$payment_intent_id?expand[]=payment_method" )
+			->will(
+				$this->returnValue(
+					json_decode( wp_json_encode( $payment_intent_mock ) )
+				)
+			);
+
+		$this->mock_gateway->process_upe_redirect_payment( $order_id, $payment_intent_id, false );
+
+		$final_order = wc_get_order( $order_id );
+		$note        = wc_get_order_notes(
+			[
+				'order_id' => $order_id,
+				'limit'    => 1,
+			]
+		)[0];
+
+		$this->assertEquals( 'processing', $final_order->get_status() );
+		$this->assertEquals( 'Visa credit card', $final_order->get_payment_method_title() );
+		$this->assertEquals( $payment_intent_id, $final_order->get_meta( '_stripe_intent_id', true ) );
+		$this->assertRegExp( '/Charge ID: ch_mock/', $note->content );
+	}
+
+	/**
+	 * Test basic checkout with saved payment method.
+	 */
+	public function test_process_payment_with_saved_method_confirms_intent_immediately() {
+		$token = $this->set_postvars_for_saved_payment_method();
+
+		$order             = WC_Helper_Order::create_order();
+		$order_id          = $order->get_id();
+		$payment_intent_id = 'pi_mock';
+		$payment_method_id = $token->get_token();
+		$customer_id       = 'cus_mock';
+
+		list( $amount, $description, $metadata ) = $this->get_order_details( $order );
+		$order->set_payment_method( WC_Stripe_UPE_Payment_Gateway::ID );
+
+		$payment_method_mock                     = self::MOCK_CARD_PAYMENT_METHOD_TEMPLATE;
+		$payment_method_mock['id']               = $payment_method_id;
+		$payment_method_mock['customer']         = $customer_id;
+		$payment_method_mock['card']['exp_year'] = intval( gmdate( 'Y' ) ) + 1;
+
+		$payment_intent_mock           = self::MOCK_CARD_PAYMENT_INTENT_TEMPLATE;
+		$payment_intent_mock['id']     = $payment_intent_id;
+		$payment_intent_mock['amount'] = $amount;
+		$payment_intent_mock['charges']['data'][0]['payment_method_details'] = $payment_method_mock;
+
+		$this->mock_gateway->expects( $this->once() )
+			->method( 'generate_payment_request' )
+			->will(
+				$this->returnValue(
+					[
+						'description' => $description,
+						'metadata'    => $metadata,
+						'capture'     => 'true',
+					]
+				)
+			);
+
+		$this->mock_gateway->expects( $this->exactly( 2 ) )
+			->method( 'stripe_request' )
+			->willReturnOnConsecutiveCalls(
+				json_decode( wp_json_encode( $payment_method_mock ) ),
+				json_decode( wp_json_encode( $payment_intent_mock ) )
+			);
```

Comment:
I also tried a bunch of different ways to mock out multiple calls to the same function and I think this was the only way I could get it to work properly. 

Also, I wanted to test the input parameters to ensure they matched what we'd expect, but for some reason the WC_Order in the third parameter had different metadata during different points of the test, so I couldn't match the expected result to the actual. Ultimately, I decided checking these inputs was not so crucial, so left them all out.